Update WC blocks e2e tests to WordPress 6.6

Aljullu opened the pull request:
### Changes proposed in this Pull Request:

This PR updates WC blocks e2e tests to work with WordPress 6.6. Most updates were related to changes in the Site Editor UI.

### How to test the changes in this Pull Request:

_(No need to test when testing the release)_

1. Make sure e2e tests pass.
2. Make sure code changes make sense.

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

Aljullu committed:
Update WC blocks e2e tests to WordPress 6.6

Aljullu committed:
Add changelog file

Aljullu committed:
Small change

Aljullu committed:
Update @wordpress/e2e-test-utils-playwright package

Aljullu committed:
Fix admin.visitSiteEditor and editor.saveSiteEditorEntities

Aljullu committed:
Update pnpm-lock.yaml

Aljullu committed:
Fix template-customization.block_theme.spec.ts

Aljullu committed:
Fix Cart, Checkout and Legacy Templates tests

Aljullu committed:
Fix tests in product-gallery.block_theme.spec.ts

Aljullu committed:
Fix tests in page-content-wrapper.block_theme.spec.ts

Aljullu committed:
Fix tests in template-customization.block_theme.spec.ts

Aljullu committed:
Fix Checkout and Local pickup tests

opr commented on plugins/woocommerce-blocks/tests/e2e/tests/templates/checkout-template.block_theme.spec.ts:
```diff
@@ -65,6 +64,7 @@ test.describe( 'Test the checkout template', () => {
 		await frontendUtils.addToCart();
 		await admin.page.goto( permalink );
 		await admin.page.locator( '#wp-admin-bar-site-editor a' ).click();
+		await admin.page.getByRole( 'button', { name: 'Get started' } ).click();
```
Is this for closing the modal? If someone is running this test locally multiple times (for debugging) would this line fail if the modal has already been dismissed in a previous run? Would it be better to check its visibility first?

WunderBart replied:
The non-deterministic state of the modal would be a side effect. Since [we reset DB for each test](https://github.com/woocommerce/woocommerce/pull/46125), side effects are no longer happening in blocks e2e tests. Therefore, we don't need to check if the modal is visible! 😄 🙌

Aljullu replied:
> Is this for closing the modal?

Yes, exactly. I added a comment in 6ee4755168af0acf9f13c63ef7487929d5437c94 to clarify a bit more what this code is doing.

> If someone is running this test locally multiple times (for debugging) would this line fail if the modal has already been dismissed in a previous run? Would it be better to check its visibility first?

My understanding is that since we are resetting the state between tests, it shouldn't make a difference, no? I made it conditional just in case, though.

Aljullu replied:
For some reason, I missed @WunderBart's comment above before publishing my reply! :sweat: 

So based on it, I removed the conditional in c4aedc0a05ecb0cd5c137519fa68e526e52aa32d.

opr replied:
Sounds good! Thanks folks :)

opr reviewed (commented):
Thanks @Aljullu this looks good to me, tests are passing as well 👌🏼 

Just one question about the get started button, once I know the answer to that I will approve.

Aljullu committed:
Make closing the welcome popup conditional

WunderBart commented on plugins/woocommerce-blocks/tests/e2e/tests/product-gallery/product-gallery.block_theme.spec.ts:
```diff
@@ -606,26 +624,21 @@ test.describe( `${ blockData.name }`, () => {
 			.getByRole( 'button', { name: 'Custom Single Product' } )
 			.click();
 
+		const editorFrame = page.frameLocator( 'iframe[name="editor-canvas"]' );
```
Can we use `editor.canvas` here?

Aljullu replied:
Good catch! Updated in e15c594a21ff149034c2e322de62ceb652868782.

WunderBart commented on plugins/woocommerce-blocks/tests/e2e/tests/templates/cart-template.block_theme.spec.ts:
```diff
@@ -27,11 +27,14 @@ test.describe( 'Test the cart template', () => {
 			.getByRole( 'button', { name: 'Cart', exact: true } )
 			.click();
-		await editor.enterEditMode();
+		await page
+			.frameLocator( 'iframe[name="editor-canvas"]' )
+			.locator( 'body' )
+			.click();
```
As above, let's use canvas here as well:

```suggestion
		await editor.canvas.locator( 'body' ).click()
```

Aljullu replied:
Updated in e15c594a21ff149034c2e322de62ceb652868782.

WunderBart commented on plugins/woocommerce-blocks/tests/e2e/tests/templates/checkout-template.block_theme.spec.ts:
```diff
@@ -29,15 +29,14 @@ test.describe( 'Test the checkout template', () => {
 			.getByRole( 'button', { name: 'Checkout', exact: true } )
 			.click();
-		await editor.enterEditMode();
+		await page
+			.frameLocator( 'iframe[name="editor-canvas"]' )
+			.locator( 'body' )
+			.click();
```
```suggestion
		await editor.canvas.locator( 'body' ).click()
```

Aljullu replied:
Updated in e15c594a21ff149034c2e322de62ceb652868782.

Aljullu committed:
Use editor.canvas instead of custom locator

WunderBart commented on plugins/woocommerce-blocks/tests/e2e/tests/templates/template-customization.block_theme.spec.ts:
```diff
@@ -149,32 +155,37 @@ test.describe( 'Template customization', () => {
 				} );
 
-				await resetButton.click();
+				// Wait until search has finished.
+				await page.waitForFunction( ( templatesLength ) => {
+					return (
+						document.querySelectorAll( '[aria-label="Actions"]' )
+							.length < templatesLength
+					);
+				}, CUSTOMIZABLE_WC_TEMPLATES.length );
```
This can be replaced with an assertion:

```suggestion
				// Wait until search has finished.
				await expect( page.getByLabel( 'Actions' ) ).toHaveCount(
					CUSTOMIZABLE_WC_TEMPLATES.length
				);
```

Aljullu replied:
What this function does is to check that the list of templates displayed on screen is "less than" the total number of WooCommerce templates. That was a simple way I found to wait until the search was performed correctly.

In the code from your suggestion, my understanding is that we would assert that the number of templates listed matches the number of WooCommerce templates, which is not exactly the same, no? :thinking:

WunderBart replied:
Ah, good catch – I saw it as `wait until less than`! In such case, I suggest providing an exact number of the expected `Actions` count when the search is done, e.g.:

```suggestion
				// Wait until search has finished.
				await expect( page.getByLabel( 'Actions' ) ).toHaveCount( 4 );
```

It should always be the same number. This way, we can be sure it's always correct and not, for example, an intermediate debounced value that would still satisfy the original condition. Does that make sense?

Aljullu replied:
The issue we have is that in most cases there is going to be only one result. Ie, if you search for "Product Search Results" it will only display one template. There is one exception where we get two results: searching for "Single Product" returns the "Single Product" template and the customized "Single Product: Belt" one. I could have hardcoded `1` and `2` depending on the test, but I thought doing a generic "wait until search has finished" was more future-proof, even though I can see how it might be more confusing.

How would you approach this case where in most tests it's `1` except for the Single Product custom template where it's `2`?

WunderBart replied:
Ah, I see. If the stability of the search is not an issue (no intermediate values), then I think we can go with `expect.poll` on this one:

```suggestion
				// Wait until search has finished.
				const searchResults = page.getByLabel( 'Actions' );
				await expect
					.poll( async () => await searchResults.count() )
					.toBeLessThan( CUSTOMIZABLE_WC_TEMPLATES.length );
```

This is the equivalent of the original condition.

Aljullu replied:
Awesome, it's definitely much cleaner. :100:

Aljullu commented:
Thanks for the reviews, folks! This is ready for another look.

Aljullu committed:
Remove closing the welcome popup conditional

Aljullu committed:
Update plugins/woocommerce-blocks/tests/e2e/tests/templates/template-customization.block_theme.spec.ts

Co-authored-by: Bart Kalisz <bartlomiej.kalisz@gmail.com>

Aljullu committed:
Update to WP 6.6 beta 3

Aljullu commented:
Just a heads-up that I rebased this PR with the latest `trunk` and updated to WordPress 6.6 beta 3 which was released yesterday.

@WunderBart @gigitux I would appreciate it if one of you could do another review. :pray:

WunderBart commented on plugins/woocommerce-blocks/tests/e2e/tests/templates/template-customization.block_theme.spec.ts:
```diff
@@ -149,32 +155,35 @@ test.describe( 'Template customization', () => {
+					.poll( async () => await searchResults.count() )
+					.toBeLessThan( CUSTOMIZABLE_WC_TEMPLATES.length );
+
+				await page.getByLabel( 'Actions' ).first().click();
```
nit
```suggestion
				await searchResults.first().click();
```

WunderBart reviewed (approved):
One more tiny suggestion. Also, if we're no longer using the `editor.enterEditMode()` util, can we remove it from the `Editor` utils?

Aljullu commented:
Thanks for taking another look, @WunderBart!

> Also, if we're no longer using the `editor.enterEditMode()` util, can we remove it from the `Editor` utils?

It has been [removed here](https://github.com/woocommerce/woocommerce/pull/48436/files#diff-5a03fcb647aaace972e20f89ccc4a8ee508a33a2d76fce077c4b1c1477219f2bL48). Is that ok or did you mean removing it from somewhere else?

Aljullu committed:
Update plugins/woocommerce-blocks/tests/e2e/tests/templates/template-customization.block_theme.spec.ts

Co-authored-by: Bart Kalisz <bartlomiej.kalisz@gmail.com>

WunderBart commented:
> Thanks for taking another look, @WunderBart!
> 
> > Also, if we're no longer using the `editor.enterEditMode()` util, can we remove it from the `Editor` utils?
> 
> It has been [removed here](https://github.com/woocommerce/woocommerce/pull/48436/files#diff-5a03fcb647aaace972e20f89ccc4a8ee508a33a2d76fce077c4b1c1477219f2bL48). Is that ok or did you mean removing it from somewhere else?

Ah, completely missed that. Thanks, that's what I meant! ❤️
